accepted/00000-http-api.md

```diff
@@ -0,0 +1,214 @@
+
+```java
+// POST https://uyuni.server/rhn/manager/api/contentmanagement/update_project/myproject
+
+post("/contentmanagement/update_project/:label", (req, res) -> {
+  // 1.Unwrap the session user, the route parameter named "label",
+  //   and the JSON object named "props" from the request body
+  // 2.Call to the `updateProject` method, passing the unwrapped arguments
+  // 3.Serialize and return the JSON result
+});
+```
+
+
+## 2. Authentication
+
+The web UI performs token-based authentication via a session cookie called `pxt-session-cookie`. Currently, the web UI login page exposes an API endpoint to support this process. This endpoint shall also be used for API authentication.
+
+### Login handshake demo
+
+```bash
+$ curl --include --header "Content-Type: application/json" \
+--data "{'login':'admin','password':'admin'}" \
+https://uyuni.dallas/rhn/manager/api/login
+
+HTTP/1.1 200 200
+...
+Set-Cookie: pxt-session-cookie=209x9fa00e2e947ace87de0743f6cf63b8b0b4bfbfca2310ce6829941fa9db947987; Max-Age=3600; Expires=Tue, 08-Feb-2022 20:05:41 GMT; Path=/; Secure; HttpOnly;HttpOnly;Secure
+...
+
+$ curl --include --cookie "pxt-session-cookie=209x9fa0..." \
+https://uyuni.server/rhn/manager/api/systems/keys
+
+HTTP/1.1 200 200
+...
+<JSON data follows>
+```
+
+
+## 3. I/O
+
+On the HTTP API side, I/O serialization is mostly handled by Spark, and the existing `GSON` library. XMLRPC uses specialized "serializer" classes to write structured output on XML. These serializer classes shall be used exclusively for the XMLRPC output.
+
+
+## 4. Exception Handling
+
+The current implementation of the business logic uses exceptions that are designed specifically for XMLRPC (see: `FaultException` and subclasses). To be able to use the business logic with both interfaces, a new, internal exception structure shall be designed to offer enough flexibility in exception handling. These exceptions are then translated to the specific interface, and processed accordingly. For HTTP endpoints, the internal exceptions can be represented by different HTTP status codes where appropriate.
+
+This is the main point that couples the business logic with a specific interface, and will require the most work to implement.
+
+
+## 5. Security
+
+The HTTP interface reuses the existing Spark framework. Therefore, we're typically not exposed to any new security issues. However, with the introduction of the new API, existing interface will be more critical for external use, so it would be beneficial to propose an approach to security testing.
+
+
+## 6. Documentation
+
+APIDOC is an extensive tool that automatically generates documentation for the API endpoints written in many languages. It can be used to generate the documentation for the HTTP API as well.
+More research is required to provide the framework for creating documentation for both interfaces with minimum duplication.
+
+
+# Unresolved Questions
```

Comment:
One topic I would like to see added is versioning. If we don't plan to do anything specific for versioning, it would also be good to write that down along with a rationale.

Reply:
I suggest using versioning in the URL (https://restfulapi.net/versioning/)